Added knowledge about how to fix deep tests (#3741)
dafny-lang/dafny: CONTRIBUTING.md

After doing these steps once, for other PRs, one only needs to re-run deep checks in (view 1)

### How to fix nightly build failures / check deep tests?

> - Create a branch that should fix the nightly build. Either an actual fix, or reverting a PR that cause the problem.

Reviewer: ```suggestion
- Create a branch that should fix the nightly build. Either an actual fix, or reverting a PR that caused the problem.
```

> - Push your branch and create a PR to merge it

Reviewer 2: ```suggestion
- Push your branch and create a PR tagged with `run-deep-tests` to merge it
```

- Go to https://github.com/dafny-lang/dafny/actions/workflows/deep-tests.yml
- Select "Run workflow..."

> - Select your branch

Reviewer 2: ```suggestion
- Select master
```

> - Wait for this new run to succeed. if necessary, update your branch.

Reviewer 2: ```suggestion
- Wait for this new run to succeed.
```

> - Go back to your PR, re-run the test "check deep tests", which should pick up the last run and now succeed.

Reviewer 2: ```suggestion
- Go back to your PR, update it, and the new CI run should pick up the last `deep-tests` run. This unblocks all tests.
```

> - Have your PR approve and merge it.

Reviewer 2: ```suggestion
```

After doing these steps once, for other PRs, one only needs to re-run deep checks in (view 1)

### How to fix nightly build failures / check deep tests?

- Create a branch (e.g. `fix-ci-<date>`) that should fix the nightly build.
  (Either an actual fix, or reverting a PR that caused the problem - look at the logs)
- Push your branch to the main Dafny repository.
- After pushing, you'll get a link to create a PR, click on it.

> - **Before clicking on the "Create PR" button**, add the label `run-deep-tests`. If you add it later, you'll have to push a new commit to trigger the deep tests.

Reviewer 2: ```suggestion
- **Before clicking on the "Create PR" button**, add the label `run-deep-tests`. If you add it later, you will have to push a new commit to trigger the deep tests.
```

- Click on the "Create PR" button.

>   The CI will run the deep tests within this PR (and for every new commit pushed to the branch). Note that only the "required" tests count.

Reviewer 2: ```suggestion
  The CI will run the deep tests within this PR (and for every new commit pushed to the branch). Note that only the "required" tests block merging.
```

- If some required tests fail, investigate and push more commits, but know that each CI run takes a lot of time, so try to fix the problem with as few commits as possible (ideally 1)

>   It is also possible you find some tests that fail randomly, in that case, feel free to relaunch them again and report to the team.

Reviewer 2: ```suggestion
  It is also possible you find some tests that fail randomly, in that case, re-run them and report to the team. Avoid using the re-run all failed jobs here, because it will re-run _all_ jobs, so select them individually instead.
```

- Have someone approve the PR and merge it (or auto-merge).
- Once the PR with the fix to the CI is merged to master, go to https://github.com/dafny-lang/dafny/actions/workflows/deep-tests.yml
- Select "Run workflow..."
- Select master
- Wait for this new run to succeed.

> - Now go back to all the PRs where check-deep-tests were failing. Relaunch the failing tests.

Reviewer 2: ```suggestion
- Now go back to all the PRs where `check-deep-tests` were failing and re-run the failing tests by updating the PRs.
```

> The new CI run should pick up the last `deep-tests` run and will make the CI item `check-deep-tests` to succeed.

Reviewer 2: ```suggestion
Subsequent CI runs should pick up the successful `deep-tests` run and make the `check-deep-tests` jobs succeed.
```
